apps/opencs/model/tools/gmstcheck.cpp

```diff
@@ -0,0 +1,96 @@
+#include "gmstcheck.hpp"
+
+#include "../world/defaultgmsts.hpp"
+
+CSMTools::GMSTCheckStage::GMSTCheckStage(const CSMWorld::IdCollection<ESM::GameSetting>& gameSettings)
+    : mGameSettings(gameSettings)
+{}
+
+int CSMTools::GMSTCheckStage::setup()
+{
+    return mGameSettings.getSize();
+}
+
+void CSMTools::GMSTCheckStage::perform(int stage, CSMDoc::Messages& messages)
+{
+    const CSMWorld::Record<ESM::GameSetting>& record = mGameSettings.getRecord (stage);
+    
+    if (record.isDeleted())
+        return;
+    
+    const ESM::GameSetting& gmst = record.get();
+    
+    CSMWorld::UniversalId id (CSMWorld::UniversalId::Type_Gmst, gmst.mId);
+    
+    // Test for empty string
+    if (gmst.mValue.getType() == ESM::VT_String && gmst.mValue.getString().empty())
+        messages.add(id, gmst.mId + " is an empty string", "", CSMDoc::Message::Severity_Warning);
+    
+    // Checking type and limits
+    // optimization - compare it to lists based on naming convention (f-float,i-int,s-string)
+    if (gmst.mId.empty())
+        messages.add(id, "Game Setting has no name", "", CSMDoc::Message::Severity_SeriousError);
+    else if (gmst.mId[0] == 'f')
+    {
+        size_t floatCount = CSMWorld::DefaultGMSTs::getFloatCount();
+        for (size_t i = 0; i < floatCount; ++i)
+        {
+            if (gmst.mId == CSMWorld::DefaultGMSTs::getFloatName(i))
+            {
+                if (gmst.mValue.getType() != ESM::VT_Float)
+                    messages.add(id, gmst.mId + " has an invalid type; it should be a float", 
+                                 "Change the GMST type to a float", CSMDoc::Message::Severity_Error);
```

Comment:
I don't think we are generally giving suggestion in error messages how to fix the error. Let's try to be consistent here. Also, the wording could be improved. The type isn't really invalid. It is just incorrect.